### PR TITLE
PMM-6585 Always expose mongodb_up metric

### PR DESCRIPTION
**Pull request by percona-csalguero**
Created a new exporter named `generalCollector` where we can put any metric that's not related to any specific collector.
Also, this collector cannot be disabled so, the `mongodb_up` metric is going to be always exposed.
Fixed error from merging connection strategy and topology labels that made tests to fail. 

FB: https://github.com/Percona-Lab/pmm-submodules/pull/1097

### CODE REVIEW

File: exporter/exporter.go [hunk 1]
```diff
@@ -96,14 +94,27 @@ func New(opts *Opts) (*Exporter, error) {
 func (e *Exporter) makeRegistry(ctx context.Context, client *mongo.Client) *prometheus.Registry {
 	// TODO: use NewPedanticRegistry when mongodb_exporter code fulfils its requirements (https://jira.percona.com/browse/PMM-6630).
 	registry := prometheus.NewRegistry()
+
+	ti, err := newTopologyInfo(context.TODO(), client)
```
[the remark this answers is not in the document]
Comment (askomorokhov) [reply exporter/exporter.go:98]: And as an idea.
If we are gathering topology on each request now, maybe we can simplify topologyInfo a few methods and simply pass labels to collectors (but not a whole instance of topologyInfo)?

**Review by askomorokhov — COMMENTED** [review]
Overall looks good. Just two comments

File: exporter/exporter.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply exporter/exporter.go:98]: Yeah, ctx is a must there

File: exporter/exporter.go [hunk 2]
```diff
@@ -96,14 +94,27 @@ func New(opts *Opts) (*Exporter, error) {
...
+
 	if len(e.opts.CollStatsCollections) > 0 {
 		cc := collstatsCollector{
 			ctx:            ctx,
 			client:         client,
 			collections:    e.opts.CollStatsCollections,
 			compatibleMode: e.opts.CompatibleMode,
 			logger:         e.opts.Logger,
-			topologyInfo:   e.topologyInfo,
+			topologyInfo:   ti,
```
Comment (AlekSi) [inline exporter/exporter.go:117]: Can those collectors handle `topologyInfo == nil`?

File: exporter/general_collector_test.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		ctx:    ctx,
+		client: client,
+	}
+
+	// The last \n at the end of this string is important
+	expected := strings.NewReader(`
+# HELP mongodb_up Whether MongoDB is up.
+# TYPE mongodb_up gauge
+mongodb_up 1` + "\n")
+	// TODO: use NewPedanticRegistry when mongodb_exporter code fulfils its requirements (https://jira.percona.com/browse/PMM-6630).
```
[the remark this answers is not in the document]
Comment (AlekSi) [reply exporter/general_collector_test.go:48]: I _think_ there we should be able to use `NewPedanticRegistry` without problems – we always return one consistent metric

File: internal/tu/testutils.go [hunk 4]
```diff
@@ -72,8 +71,9 @@ func TestClient(ctx context.Context, port string, t *testing.T) *mongo.Client {
 	require.NoError(t, err)
 
 	t.Cleanup(func() {
-		err := client.Disconnect(ctx)
-		assert.NoError(t, err)
+		// In some cases, we manually disconnect the client so, don't check for errors,
+		// it might be already disconnected.
+		client.Disconnect(ctx) //nolint
```
Comment (AlekSi) [inline internal/tu/testutils.go:76]: Let's be specific if we can. I guess `//nolint:errcheck` should work

File: internal/tu/testutils.go [hunk 5]
```diff
@@ -72,8 +71,9 @@ func TestClient(ctx context.Context, port string, t *testing.T) *mongo.Client {
 	require.NoError(t, err)
 
 	t.Cleanup(func() {
-		err := client.Disconnect(ctx)
-		assert.NoError(t, err)
+		// In some cases, we manually disconnect the client so, don't check for errors,
+		// it might be already disconnected.
```
Comment (AlekSi) [inline internal/tu/testutils.go:75]: 👍

File: exporter/exporter.go [hunk 6]
```diff
@@ -87,15 +89,33 @@ func New(opts *Opts) (*Exporter, error) {
...
 func (e *Exporter) makeRegistry(ctx context.Context, client *mongo.Client) *prometheus.Registry {
-	// TODO: use NewPedanticRegistry when mongodb_exporter code fulfils its requirements (https://jira.percona.com/browse/PMM-6630).
-	registry := prometheus.NewRegistry()
+	registry := prometheus.NewPedanticRegistry()
+
+	e.lock.Lock()
+	if e.topologyInfo == nil {
+		var err error
+
+		if e.topologyInfo, err = newTopologyInfo(ctx, client); err != nil {
```
Comment (askomorokhov) [inline exporter/exporter.go:106]: e.topologyInfo` can be local variable as we not use it outside of the current method.
In this case, `e.lock` will obsolete as well.

**Review by askomorokhov — APPROVED** [review]

**Review by BupycHuk — CHANGES_REQUESTED** [review]
please revert all changes related to topologyInfo

**Comment by percona-csalguero** [comment]
why?

**Comment by percona-csalguero** [comment]
Closing in favor of https://github.com/percona/mongodb_exporter/pull/208

**Pull request closed without merging**